# Staff-Graded Recap Matrices

### Part (a)


In [1]:
import numpy as np
from scipy.sparse import csr_matrix
import time
import pandas as pd

data = np.genfromtxt('TexasCityDistanceMatrix.csv', delimiter=',')
key = pd.read_csv('TexasCityKey.csv', encoding = "ISO-8859-1")

# Drop the first row and column
matrix = data[1:, 1:]

# Find the maximum value ignoring NaN
max_value = np.nanmax(matrix)
print(max_value)

99.99971353


The maximum value in the matrix is 99.9997. This seems to indicate that the NA values in the matrix represent values of 100 or greater.

### Part (b)


In [2]:
# Note that the row and column headers have been removed, and thus Katy is the 750th row in the matrix
# in its current form

# Specify the row index
row_index = 749

# Find the column index of the lowest non-zero value that is not NaN in the specified row
row_values = matrix[row_index]
nonzero_indices = np.where((row_values != 0) & (~np.isnan(row_values)))[0]

min_index = np.argmin(row_values[nonzero_indices])
column_index = nonzero_indices[min_index]

print(f"Column index of the lowest non-zero value in row {row_index} is {column_index}")
print(key.iloc[column_index,1])


# Find locations within 25 miles of Flower Mound
# Flower Mound is the 507th row in the matrix
# Find row of interest
row_values = matrix[506, :]

# Remove zeroes
nonzero_values = row_values[row_values != 0]

# Find values less than 25
values = nonzero_values[nonzero_values <= 25]
print(len(values))

Column index of the lowest non-zero value in row 749 is 276
Cinco Ranch CDP
75


The closest location to Katy is Cinco Ranch.

There are 75 locations within 25 miles of Flower Mound.

### Part (c)


In [14]:
start = time.time()
np.sum(matrix)
end = time.time()
print(end - start)

0.004388093948364258


On my machine and depending on the run, this operation takes approximately 0.004 seconds.

### Part (d)


In [22]:
# Set values greater than 50 to zero
mod_matrix = matrix
mod_matrix[mod_matrix > 50] = 0

# Set NAs to zero
mod_matrix = np.nan_to_num(mod_matrix)

start = time.time()
np.sum(mod_matrix)
end = time.time()
print(end - start)

0.0020589828491210938


On my machine and depending on the run, this operation takes approximately 0.002 seconds.

### Part (e)

In [40]:
sparse_matrix = csr_matrix(matrix)

start = time.time()
np.sum(sparse_matrix)
end = time.time()
print(end - start)

0.006075859069824219


### Part (f)
Knowing that the matrix is symmetric and has zeroes on the diagonal, we could save space by storing the matrix in some data structure other than a NumPy array. For the left triangle of the matrix, for non-zero elements, we could store the element value and row-column index pair. It would then be implicit that all elements mirrored across the diagonal are equal. In other words, element (j,i) equals element (i,j). It would also be implicit that index pairs not specifically associated with a non-zero element value are zero. In essence, we do not need to store these zero elements at all, as they can be reintroduced at a later time if needed.